pkg/endpointmanager/cell.go

```diff
@@ -0,0 +1,199 @@
+	// only be done once, globally, otherwise the metrics library will panic.
+	InitMetrics()
+
+	// Subscribe to endpoint events.
+	Subscribe(s Subscriber)
+
+	// Unsubscribe from endpoint events.
+	Unsubscribe(s Subscriber)
+
+	// UpdatePolicyMaps returns a WaitGroup which is signaled upon once all endpoints
+	// have had their PolicyMaps updated against the Endpoint's desired policy state.
+	//
+	// Endpoints will wait on the 'notifyWg' parameter before updating policy maps.
+	UpdatePolicyMaps(ctx context.Context, notifyWg *sync.WaitGroup) *sync.WaitGroup
+
+	// RegenerateAllEndpoints calls a setState for each endpoint and
+	// regenerates if state transaction is valid. During this process, the endpoint
+	// list is locked and cannot be modified.
+	// Returns a waiting group that can be used to know when all the endpoints are
+	// regenerated.
+	RegenerateAllEndpoints(regenMetadata *regeneration.ExternalRegenerationMetadata) *sync.WaitGroup
+
+	// WaitForEndpointsAtPolicyRev waits for all endpoints which existed at the time
+	// this function is called to be at a given policy revision.
+	// New endpoints appearing while waiting are ignored.
+	WaitForEndpointsAtPolicyRev(ctx context.Context, rev uint64) error
+
+	// OverrideEndpointOpts applies the given options to all endpoints.
+	OverrideEndpointOpts(om option.OptionMap)
+
+	// InitHostEndpointLabels initializes the host endpoint's labels with the
+	// node's known labels.
+	InitHostEndpointLabels(ctx context.Context)
+
+	// GetPolicyEndpoints returns a map of all endpoints present in endpoint
+	// manager as policy.Endpoint interface set for the map key.
+	GetPolicyEndpoints() map[policy.Endpoint]struct{}
+
+	// HasGlobalCT returns true if the endpoints have a global CT, false otherwise.
+	HasGlobalCT() bool
+
+	// CallbackForEndpointsAtPolicyRev registers a callback on all endpoints that
+	// exist when invoked. It is similar to WaitForEndpointsAtPolicyRevision but
+	// each endpoint that reaches the desired revision calls 'done' independently.
+	// The provided callback should not block and generally be lightweight.
+	CallbackForEndpointsAtPolicyRev(ctx context.Context, rev uint64, done func(time.Time)) error
+}
+
+var (
+	_ EndpointsLookup = &endpointManager{}
+	_ EndpointsModify = &endpointManager{}
+	_ EndpointManager = &endpointManager{}
+)
+
+type endpointManagerParams struct {
+	cell.In
+
+	Lifecycle hive.Lifecycle
+	Config    EndpointManagerConfig
+	Clientset client.Clientset
+}
+
+type endpointManagerOut struct {
+	cell.Out
+
+	Lookup  EndpointsLookup
+	Modify  EndpointsModify
+	Manager EndpointManager
+}
+
+func newDefaultEndpointManager(p endpointManagerParams) endpointManagerOut {
+	checker := endpoint.CheckHealth
+
+	mgr := New(&watchers.EndpointSynchronizer{Clientset: p.Clientset})
+	if p.Config.EndpointGCInterval > 0 {
+		ctx, cancel := context.WithCancel(context.Background())
+		mgr = mgr.WithPeriodicEndpointGC(ctx, checker, p.Config.EndpointGCInterval)
```

Comment:
`newDefaultEndpointManager` is basically what `WithDefaultEndpointManager` used to be. The latter used to take a root ctx from the Daemon [1], but the former doesn't allow for that. Can we make this code path accept a ctx as to keep the same behavior as prior to this PR?

Reply:
With hive/cells there will be no root context anymore (which was problematic as it provided no mechanism to wait for things to finish). Each module is supposed to manage their own context (or similar mechanism) whose lifecycle is determined by the Start and Stop hooks.

Reply:
Ah, thanks for the explanation!